### ISSUE TITLE
IF ... == ... raises exception (?)

### ISSUE THREAD

**Issue by vcxsd** [robhagemans/pcbasic#5]
I was tinkering and typed something like this into the command line (i.e. "immediate mode" or something, I guess):

```
if "STRING ONE" == "STRING TWO" then print "UH OH"
```

I got this (copied from the console, not the PCBASIC window, but it's the same thing really):

```
ERROR: Traceback (most recent call last):
  File "build/pcbasic/run.py", line 66, in run_once
  File "build/pcbasic/statements.py", line 114, in parse_statement
  File "build/pcbasic/statements.py", line 1563, in exec_if
  File "build/pcbasic/expressions.py", line 117, in parse_expression
KeyError: '\xe7\xe7'
```

I have no idea what this means (that is, I don't know how PCBASIC works on the inside, so it's just a cryptic KeyError to me.) I assumed that meant you might want to hear about it.

I've tested with a few other immediate mode IF statements and equality (empty strings, arbitrary strings, numbers...) and PRINT, same result.

It seems like using a single `=` actually works as expected. So the error is just down to my not being so familiar with BASIC. Still I would expect a `Syntax error` or something, not this, so maybe it's worth hearing about. I don't know. Doing this in GWBASIC in DOSBOX (yes... I have *no good reason* to do this, I know...) does seem to give an ordinary syntax error.

I'm using the stable release version (0.15*.) I'm not using the 0.16 version because it froze up my computer multiple times when I tried.

**Comment by robhagemans** [comment]
Thanks for the information! That's a bug, it should indeed give a `Syntax error`. (As you've noticed, BASIC uses single `=` for both comparison and assigment operations). I'll fix the bug for the next stable release. 

This buf is probably in the 16.06 version as well so that wouldn't have helped you, but I'm interested to know more about the issues you're experiencing with that version. What operating system do you use? What exactly happens when it freezes up your computer?

**Comment by vcxsd** [comment]
I've been running Arch Linux on this computer for some time now. It's just an old laptop, 2GB of RAM, but normally works fine. So far as I can remember v0.16 worked for a minute or two, but then the entire system started to lag (as in, mouse movement lagging, even, I think?) and I found it less frustrating to manually turn the machine off and start it again rather than trying to kill the interpreter somehow, because things just weren't responding. This might have happened twice, but after that I definitely decided I wasn't interested in trying it a third time.

I'm not sure what that was. I would presume something that managed to consume all the system's resources at once. Somehow. I don't think that could be done with just a simple infinite loop, though, and I wasn't looking at a system monitor while it happened. If I had to take a guess, I think I'd guess that it might have been something to do with SDL2, as that seemed to be described as the major change in version .16 and I had installed pysdl2 somewhere to get the graphical interface working, but not pygame. I don't think I had every single dependency installed, either. I don't know anything about how SDL2 or pySDL2 works internally, though, so this is mostly just a shot in the dark.

If you have a particular debug mode you'd like me to try, or just reinstall .16 and run it, watch a system monitor and see if the problem occurs again, I can do that. I'm not sure how far I want to go, though; it would be rather inconvenient to do this a lot if it does keep happening.

**Comment by robhagemans** [comment]
Thanks! The one thing that may be quite useful to try is to run 16.06 with PyGame rather than SDL2 through the command `pcbasic --interface=pygame` and see if the problem also arises. If not, at least we know it's related to SDL2 or the code that interfaces with it.

I've fixed the `==` bug in commit b483375.

**Comment by robhagemans** [comment]
Some other things that would be useful to know is if the freezes also occur if you use `pcbasic --scaling=native` or `pcbasic --scaling=crisp` (if not, we'd be able to narrow the problem down to SDL2_gfx) and if you use `pcbasic --nosound` (to see if we can blame the sound system). 
It would also be very useful to have `top` open in another terminal to see which process (if any) is hogging the CPU. It could simply be `pcbasic` but it may also be the case that, for example, pcbasic is causing PulseAudio (if that's what Arch uses) or X or your window manager or something to run wild.
Speaking of which, which window manager/desktop environment and sound system do you use on Arch?

(I can see how it would get a bit annoying to freeze up your computer and reboot a lot of times, so I understand if it is too inconvenient to try much, but if you are looking to try out a few things the above are things that come to mind...)

Cheers
Rob

**Comment by vcxsd** [comment]
I had a bit of unexpected time this morning and did some testing. (Why not...) It seems like the problem is a memory leak -- watching the system monitor, it didn't use an excessive amount of CPU, but the RAM used by the process (and overall) increased until the system froze. This doesn't happen when using `--interface=pygame`, but does happen with `--scaling=crisp` and `--nosound` (though I didn't let it go so far as to freeze my machine again when testing those.) The increase in memory usage is on the order of at least a few megabytes every time `htop` refreshes, so far as I can tell.

If you want my pet theory at this point, it's that something, somewhere, is allocating new memory for each frame, or something similar, rather than re-using the old memory. If this generally works for other people I would imagine it's something that changed in (py)SDL2, maybe how the Python side interfaces with the native-code SDL2 side? Arch Linux tends to serve the latest, most bleeding-edge versions of software packages (though I think I used Python's packaging mechanism to install the SDL2 bindings), so it might just be some weird interaction between versions, or something. Or it might not. The problem could still be somewhere in these versions of (py)SDL2 rather than in PCBASIC itself, though.

In case it's helpful, my package manager says this about what versions of sdl2-related packages are installed on my system:

```
local/sdl2 2.0.4-2
    A library for portable low-level access to a video framebuffer, audio
    output, mouse, and keyboard (Version 2)
local/sdl2_gfx 1:1.0.1-1
    SDL Graphic Primitives (Version 2)
local/sdl2_image 2.0.1-1
    A simple library to load images of various formats as SDL surfaces
    (Version 2)
local/sdl2_mixer 2.0.1-1
    A simple multi-channel audio mixer (Version 2)
```

**Comment by robhagemans** [comment]
Awesome, thanks for the research! I'll have a look if I can find the memory leak. It may very well be that my code is forgetting to free some memory it has allocated on a frame cycle - if so, I should be able to see the increase in memory usage on my own setup; I may simply not have noticed with 8GB RAM. If the leak is in a particular version or build of SDL2 then things are of course more complicated, but I'll check the easy stuff first. Will let you know!

**Comment by robhagemans** [comment]
Your theory was spot on - a surface was being created and not freed on each cycle. Fixed in commit 1d9a038. With this change memory usage now appears stable on my machine. Does it solve the issue for you?

Cheers
Rob

**Comment by vcxsd** [comment]
Hey, that does seem to fix it! Neat...

I did have to comment out a `from . import ansipipe` to get the github repository to run, which I didn't notice in any of the instructions -- but then I more or less just skimmed some of them, so... eh.

**Comment by robhagemans** [comment]
Cool, I'll consider that one fixed then!

It's OK to comment out `ansipipe`, it does literally nothing on Linux. It basically exists because of Windows being a total pain about anything console-related. The reason it didn't work out of the box is probably because `ansipipe` is included as a submodule and git doesn't automatically pull those in unless you use `git clone --recursive`. Which `INSTALL.md` does suggest to do, but it's easy to miss... 
I just tried to follow the step by step instructions and noted some things I've overlooked, so I'll rewrite that section a bit anyway.
